feat: Support submissions list filtering

dspetrov opened the pull request:
# Support submissions list filtering

## Description
Implement general support of filtering in Endatix API following #248. Make the changes for Submissions List endpoint to support filtering.
**Note** It is ready for review, Unit tests are still not added but will be.

## Testing
Using Swagger is on way to test, e.g.:
![image](https://github.com/user-attachments/assets/0a9dec89-ea18-4d50-a5ba-ebef9d29fd6a)
Manual executions with the filters in the query string also will work and any tool for web requests will be helpful.

## Related Issues
closes #250

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the style guidelines of this project

PeterMarinov commented on src/Endatix.Core/Specifications/Parameters/FilterCriterion.cs:
```diff
@@ -0,0 +1,96 @@
+                return operatorIndex == 0 || filterExpression[..operatorIndex].All(c => char.IsLetterOrDigit(c));
+            });
+
+        Guard.Against.Null(
+            @operator,
+            nameof(filterExpression),
+            $"Invalid filter operator. Valid operators are: {string.Join(", ", _operatorMap.Keys)}");
+
+        var parts = filterExpression.Split(@operator, 2);
+        Guard.Against.InvalidInput(
+            parts,
+            nameof(filterExpression),
+            p => p.Length == 2,
+            "Invalid filter format. Expected format: field{operator}value where {operator} is one of ':', '!:', '>', '<', '>:', '<:'");
+
+        Field = Guard.Against.NullOrWhiteSpace(parts[0].Trim(), nameof(Field));
+        Operator = _operatorMap[@operator];
+
+        var values = parts[1].Split(',').Select(v => v.Trim()).ToList();
+        Guard.Against.NullOrEmpty(
+            values,
+            nameof(filterExpression),
+            "Filter must have at least one value");
+        Guard.Against.InvalidInput(
+            values,
+            nameof(filterExpression),
+            v => !v.Any(string.IsNullOrWhiteSpace),
+            "Filter values cannot be empty or whitespace");
```
Upon checking the rest of the PR I assume your idea is that the validator should do the heavy lifting, so exceptions are less likely. This is good enough for me even though I prefer we don't throw and return the 400 when filter is malformed or not possible. No actions needed

PeterMarinov reviewed (dismissed):
I like the PR. I have approved and added directions for discussions. Since this is foundational work with high future impact, I feel it's good to discuss items and log next improvements once we agree on common approach.

dspetrov commented on tests/Endatix.Core.Tests/UseCases/Submissions/ListByFormId/ListByFormIdHandlerTests.cs:
```diff
@@ -48,7 +48,7 @@ public async Task Handle_ValidRequest_ReturnsSubmissions()
             new("{ }", 1, 2) { Id = 4 }
         };
-        var request = new ListByFormIdQuery(1, 1, 10);
+        var request = new ListByFormIdQuery(1, 1, 10, new List<string>());
```
Yes, thanks. I will.

dspetrov commented on src/Endatix.Core/Specifications/SubmissionsByFormIdSpec.cs:
Do you mean to fix the warnings in SubmissionsByFormIdSpec.cs?
I don's see any warnings. I've seen some proposals from the IDE about the XML comments. I've changed them and there aren't any proposals anymore.

dspetrov commented on src/Endatix.Api/Endpoints/Submissions/ListByFormId.ListByFormIdValidator.cs:
```diff
@@ -1,13 +1,29 @@
 
 public class ListByFormIdValidator : Validator<ListByFormIdRequest>
 {
+    private static readonly Dictionary<string, Type> _filterableFields = new()
```
Yes, it's good to have in mind that defining the fields will evolve in the future. Here these are only for validation purposes. It's hard now to say which is the right place to get them defined (e.g. application, tenant, etc) as this will largely depend on the business requirements and needs. I would say we start with the most basic - the type has a set of filterable columns it is the direct owner and we evolve to more flexible approaches depending on the needs.

dspetrov commented on src/Endatix.Core/Specifications/SpecificationExtensions.cs:
```diff
@@ -30,4 +34,110 @@ public static ISpecificationBuilder<TEntity> Paginate<TEntity>(this ISpecificati
+            return query;
+        }
+
+        var entityType = typeof(TEntity);
+        var propertyInfo = entityType.GetProperty(
+            filter.Field, 
+            BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);
+        
+        Guard.Against.Null(
+            propertyInfo,
+            nameof(filter.Field), 
+            $"Property '{filter.Field}' does not exist on type '{entityType.Name}'");
+
```
I agree the reflection has its drawback. I'm not sure however, how much it is used around our code base and especially in the libraries we use. I think it's good to have some analysis and let the numbers speak before we put efforts for optimizations. For sure the reflection is slow but who knows it may turn out that compared to a single DB request is almost nothing and optimizing it could improve very little... just an example.
I don't really understand your example from the code you share.
Generally I would suggest we plan and start doing performance testing so we can analyze where are the bottlenecks, then really focus our efforts on what mostly slows our system.
As for caching - this really sounds good and will be an easy way to considerably lower the reflection calls. If you agree, let's start with this step. I can add a task for caching here.

PeterMarinov replied:
I have seen apps crashing because of the reflection overhead and hence me being proactive about it. Fully agree on the stress testing instead of blind premature optimizations - this is the way!

For the caching task, yes - please add it. It is an easy win here especially since we don't need cache invalidation (fields don't change unless the code change). For the task, please mark it with  [:zap: performance](https://github.com/endatix/endatix/labels/%3Azap%3A%20performance) and [tech-debt](https://github.com/endatix/endatix/labels/tech-debt).

dspetrov replied:
Here it is - https://github.com/endatix/endatix/issues/301

dspetrov commented:
I've added the tests and it needs a new approval.
While adding the tests I made a few small changes:
- enhanced some of the checks in FilterCriterion
- changed doc comments in a few places
- moved SpecificationExtensions to folder common, which needed some namespace changes
- Extracted the filtering logic from SpecificationExtensions to SpecificationHelper so it can be tested (it's hard to compare in the commit as it's in two files but it's the same code, just moved)
All the above can be seen in the last commit and although includes a number of files, it doesn't change anything in how the implementation works in practice.

@PeterMarinov, please have a look and approve again if any questions don't appear.
